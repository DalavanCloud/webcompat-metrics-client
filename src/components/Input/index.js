import React from "react";
import PropTypes from "prop-types";

import classes from "./styles.css";

class Input extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      value: props.defaultValue,
    };

    this.handleChange = this.handleChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.defaultValue !== this.props.defaultValue) {
      this.setState({ value: nextProps.defaultValue });
    }
  }

  handleChange(e) {
    this.setState({
      value: e.target.value,
    });

    this.props.onChange(e);
  }

  render() {
    const { required, style, ...attrs } = this.props;

    delete attrs.value;
    delete attrs.defaultValue;

    return (
      <input
        className={classes.component}
        style={style}
        {...(required ? { required: true } : undefined)}
        value={this.state.value}
        {...attrs}
        onChange={this.handleChange}
      />
    );
  }
}

Input.propTypes = {
  style: PropTypes.object,
  type: PropTypes.string,
  name: PropTypes.string.isRequired,
  defaultValue: PropTypes.string,
  required: PropTypes.bool,
  onChange: PropTypes.func,
};

Input.defaultProps = {
  style: {},
  type: "text",
  defaultValue: "",
  required: false,
  onChange() {},
};

export default Input;
