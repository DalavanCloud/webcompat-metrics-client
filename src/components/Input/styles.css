/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

:root {
  --fontSize: var(--fontSize);
  --fontFamily: var(--fontFamily);
}

.component {
  display: inline-flex;
  font-size: var(--fontSize);
  font-family: var(--fontFamily);
  font-weight: 300;
  padding: 0.6em;
  margin: 0;
  line-height: normal;
  vertical-align: middle;
  background-color: #fff;
  color: #404040;
  transition: background-color 0.3s ease 0s;
  border: solid 1px #dee5e7;
}
