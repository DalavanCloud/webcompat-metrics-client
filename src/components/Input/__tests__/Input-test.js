/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from "react";
import renderer from "react-test-renderer";
import sinon from "sinon";
import { shallow } from "enzyme";

import Input from "..";

it("renders Input default correctly", () => {
  const tree = renderer.create(<Input name="input" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders Input with defaultValue", () => {
  const tree = renderer.create(<Input name="input" value="my name" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders Input required", () => {
  const tree = renderer.create(<Input name="input" required={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders Input with style injected", () => {
  const tree = renderer
    .create(<Input name="input" style={{ color: "red" }} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders Input: simulates onChange events", () => {
  const onChange = sinon.spy();
  const event = { target: { name: "input", value: "test" } };
  const wrapper = shallow(<Input name="input" onChange={onChange} />);
  wrapper.find("input").simulate("change", event);
  expect(onChange.calledOnce).toEqual(true);
});

it("renders Input with a prop update", () => {
  const wrapper = shallow(<Input name="input" />);
  wrapper.setProps({ value: "new value" });
  expect(wrapper.props().value).toEqual("new value");
  wrapper.setProps({ value: "new value" });
});
