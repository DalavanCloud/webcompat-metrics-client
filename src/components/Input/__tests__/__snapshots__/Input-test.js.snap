// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Input default correctly 1`] = `
<input
  className="component"
  name="input"
  onChange={[Function]}
  type="text"
  value=""
/>
`;

exports[`renders Input required 1`] = `
<input
  className="component"
  name="input"
  onChange={[Function]}
  required={true}
  type="text"
  value=""
/>
`;

exports[`renders Input with defaultValue 1`] = `
<input
  className="component"
  name="input"
  onChange={[Function]}
  type="text"
  value=""
/>
`;

exports[`renders Input with style injected 1`] = `
<input
  className="component"
  name="input"
  onChange={[Function]}
  style={
    Object {
      "color": "red",
    }
  }
  type="text"
  value=""
/>
`;
